### ISSUE TITLE
Kubernetes Access Logs - Final Field In Config Not Dropped

### ISSUE THREAD

**Issue by WillJCJ** [containous/traefik#3841]
### Do you want to request a *feature* or report a *bug*?

Bug

### What did you do?

I enabled access logging on my Tæfik ingress controller. I do not want to log `Username`, `Hostname`, `ClientUsername` and `ClientHost`.

To do this, I passed the command line arguments:

```console
--accessLog.fields.defaultMode="keep"
--accessLog.fields.names="Username=drop Hostname=drop ClientUsername=drop ClientHost=drop"
```

The same thing occurs with headers:

```console
--accessLog.fields.headers.defaultMode=keep
--accessLog.fields.headers.names="Authorization=redact Cookie=redact Set-Cookie=redact X-Csrftoken=redact X-Forwarded-For=redact"
```

### What did you expect to see?

I expected `ClientHost` to not appear in the logs.

I expected to see `"request_X-Forwarded-For":"REDACTED"` in the JSON logs.

### What did you see instead?

`ClientHost` was present in the access logs.
`request_X-Forwarded-For` had an IP address.

All other fields were dropped and redacted.

### Output of `traefik version`: (_What version of Traefik are you using?_)

Noticed the problem on `1.6.4`

Recreated the issue on `1.6.6`

Recreated the issue on `v1.7.0-rc3`

### What is your environment & configuration (arguments, toml, provider, platform, ...)?

The environment is a Kops cluster in AWS running `v1.9.7` of Kubernetes.

I'm not using a TOML file, just the command line args in a Kubernetes `yaml` file.

Below is my `yaml` file for the deployment.

```yaml
---
apiVersion: apps/v1beta2
kind: Deployment
metadata:
  name: traefik-ingress-lb
  labels:
    app: traefik-ingress-lb
    tier: proxy
spec:
  replicas: 2
  selector:
    matchLabels:
      app: traefik-ingress-lb
      tier: proxy
  template:
    metadata:
      labels:
        app: traefik-ingress-lb
        tier: proxy
      annotations:
        scheduler.alpha.kubernetes.io/critical-pod: ''
        scheduler.alpha.kubernetes.io/tolerations: '[{"key":"CriticalAddonsOnly", "operator":"Exists"}]'
    spec:
      serviceAccountName: traefik-ingress
      terminationGracePeriodSeconds: 60
      containers:
      - image: traefik:v1.7
        name: traefik-ingress-lb

        ports:
        - name: http
          containerPort: 80
        - name: https
          containerPort: 443
        - name: admin
          containerPort: 8080

        args:
        - --traefikLog.format=json
        - --accessLog.format=json
        - --accessLog.fields.defaultMode="keep"
        - --accessLog.fields.names="Username=drop Hostname=drop ClientUsername=drop ClientHost=drop"
        - --accessLog.fields.headers.defaultMode=keep
        - --accessLog.fields.headers.names="Authorization=redact Cookie=redact Set-Cookie=redact X-Csrftoken=redact X-Forwarded-For=redact"
        - --api
        - --api.dashboard
        - --api.statistics
        - --kubernetes
        - --kubernetes.ingressClass=traefik-external
        - --kubernetes.namespaces=app
        - --metrics.prometheus
        - --logLevel=INFO
        - --entryPoints=Name:http Address::80
        - --entryPoints=Name:https Address::443 TLS
        - --defaultentrypoints=http

        resources:
          requests:
            cpu: "100m"
            memory: "20Mi"
          limits:
            cpu: "150m"
            memory: "40Mi"

        readinessProbe:
          tcpSocket:
            port: 80
          failureThreshold: 1
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 2
        livenessProbe:
          tcpSocket:
            port: 80
          failureThreshold: 3
          initialDelaySeconds: 10
          periodSeconds: 10
          successThreshold: 1
          timeoutSeconds: 2
```

#### Logs
Each request makes an access log for the pod, containing

`"request_X-Forwarded-For":"12.34.45.67"` (Should be REDACTED)
`"ClientHost":"12.34.56.78"` (Shouldn't be here)

### Fix

A dirty fix is to add an extra field to each command line arg:

```console
--accessLog.fields.names="Username=drop Hostname=drop ClientUsername=drop ClientHost=drop FakeField=drop"
--accessLog.fields.headers.names="Authorization=redact Cookie=redact Set-Cookie=redact X-Csrftoken=redact X-Forwarded-For=redact X-Clacks-Overhead=keep"
```

This then makes the logs function as expected.

### Recreation

I can only re-create this bug with the exact `yaml` file I put above.

I tried just using `docker run`  with the same args. The log fields were removed and redacted as expected.

I tried running træfik with the same args by itself in `minikube` cluster on my laptop and again, the fields were dropped and redacted.

**Comment by Ullaakut** [comment]
Issue reproduced locally on the `1.7`. Looking for the issue in the access log middleware.

**Comment by Ullaakut** [comment]
I think I see what the problem is, and I'm going to try to fix it. In the meantime, a workaround that works for me is to simply remove the double quotes like this:

```
        - --accessLog.fields.names=Username=drop Hostname=drop ClientUsername=drop ClientHost=drop
        - --accessLog.fields.headers.names=Authorization=redact Cookie=redact Set-Cookie=redact X-Csrftoken=redact X-Forwarded-For=redact
```

It seems the issue is that because of the YAML format, the parser gets escaped characters when there are double quotes in the line, and that's what is making it ignore the last key-value pair.

Also, I found another issue while experimenting with this, which is that if we add a space at the end without providing another key/value pair, the parser crashes:

```
panic: runtime error: index out of range

goroutine 1 [running]:
github.com/containous/traefik/types.(*FieldHeaderNames).Set(0xc0000a6020, 0xc0001f00b1, 0x61, 0x1e, 0x2)
	/go/src/github.com/containous/traefik/types/logs.go:130 +0x153
github.com/containous/traefik/vendor/github.com/ogier/pflag.(*FlagSet).setFlag(0xc0006a7280, 0xc000357040, 0xc0001f00b1, 0x61, 0xc0001f0090, 0x82, 0x2, 0x2)
	/go/src/github.com/containous/traefik/vendor/github.com/ogier/pflag/flag.go:464 +0x63
github.com/containous/traefik/vendor/github.com/ogier/pflag.(*FlagSet).parseArgs(0xc0006a7280, 0xc0000ac4e0, 0xc, 0xb, 0x0, 0x11)
	/go/src/github.com/containous/traefik/vendor/github.com/ogier/pflag/flag.go:516 +0xc28
github.com/containous/traefik/vendor/github.com/ogier/pflag.(*FlagSet).Parse(0xc0006a7280, 0xc0000ac480, 0x11, 0x11, 0x11, 0x11)
	/go/src/github.com/containous/traefik/vendor/github.com/ogier/pflag/flag.go:563 +0xa1
github.com/containous/traefik/vendor/github.com/containous/flaeg.parseArgs(0xc0000ac010, 0x11, 0x11, 0xc0006a7500, 0xc0005dc060, 0x16, 0xc0006a74d0, 0x0)
	/go/src/github.com/containous/traefik/vendor/github.com/containous/flaeg/flaeg.go:139 +0x7e9
github.com/containous/traefik/vendor/github.com/containous/flaeg.LoadWithCommand(0xc0002ba120, 0xc0000ac010, 0x11, 0x11, 0xc0002a9800, 0xc0005ba140, 0x5, 0x8, 0x0, 0x0)
	/go/src/github.com/containous/traefik/vendor/github.com/containous/flaeg/flaeg.go:419 +0x35b
github.com/containous/traefik/vendor/github.com/containous/flaeg.(*Flaeg).Parse(0xc0002ba1e0, 0xc0002ba120, 0x0, 0x0, 0x5)
	/go/src/github.com/containous/traefik/vendor/github.com/containous/flaeg/flaeg.go:640 +0x7d
main.main()
	/go/src/github.com/containous/traefik/cmd/traefik/traefik.go:91 +0x1401
```

**Comment by WillJCJ** [comment]
Thanks for taking a look, I wouldn't have known where to start!
Your fix is certainly nicer than adding fake entries to the end of the list.

**Comment by traefiker** [comment]
Closed by #4113.
